src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UserNotificationServices.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    ///     Provides an implementation of <see cref="IUserNotificationServices"/> that delegates onto 
+    /// </summary>
+    [Export(typeof(IUserNotificationServices))]
+    internal class UserNotificationServices : IUserNotificationServices
+    {
+        private readonly SVsServiceProvider _serviceProvider;
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IOptionsSettings _optionsSettings;
+
+        [ImportingConstructor]
+        public UserNotificationServices(IOptionsSettings optionsSettings, SVsServiceProvider serviceProvider, IProjectThreadingService threadingService)
+        {
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+            Requires.NotNull(threadingService, nameof(threadingService));
+            Requires.NotNull(optionsSettings, nameof(optionsSettings));
+            _serviceProvider = serviceProvider;
+            _threadingService = threadingService;
+            _optionsSettings = optionsSettings;
+        }
+
+        public async Task<bool> CheckPromptAsync(string promptMessage)
+        {
+            await _threadingService.SwitchToUIThread();
+
+            var result = VsShellUtilities.ShowMessageBox(_serviceProvider, promptMessage, null, OLEMSGICON.OLEMSGICON_QUERY,
+                          OLEMSGBUTTON.OLEMSGBUTTON_YESNO, OLEMSGDEFBUTTON.OLEMSGDEFBUTTON_FIRST);
+            if (result == (int)VSConstants.MessageBoxResult.IDNO)
+            {
+                return false;
+            }
+            return true;
+        }
+
+        public async void NotifyFailureAsync(string failureMessage)
+        {
+            await _threadingService.SwitchToUIThread();
```

Comment:
We shouldn't have async voids as they aren't observable, and exceptions go unhandled.

I'm also unconvinced by this pattern of having "async" methods that are really just pumping to the UI thread. I think we should change the notification service to be synchronous and have the caller worry about switching to the UI thread. We should have such services check to make sure they are called on the UI thread, I'll do some research as CPS might already have some helper classes for types that are not free-threaded.